hudi-common/src/main/java/org/apache/hudi/common/lock/HdfsFileLockProvider.java

```diff
@@ -0,0 +1,125 @@
+ */
+public class HdfsFileLockProvider implements LockProvider<String> {
+  private static final Logger LOG = LogManager.getLogger(HdfsFileLockProvider.class);
+
+  private String lockDir;
+  private FileSystem fs;
+  protected LockConfiguration lockConfiguration;
+  private DataOutputStream dataOutputStream;
+  private String lockFile;
+
+  public HdfsFileLockProvider(final LockConfiguration lockConfiguration, final Configuration conf) {
+    this.lockConfiguration = lockConfiguration;
+    this.lockDir = lockConfiguration.getConfig().getString(LockConfiguration.FILESYSTEM_LOCK_PATH_PROP_KEY);
+    this.fs = FSUtils.getFs(lockDir, conf);
+    this.lockFile = new Path(lockDir, "lockFile").toUri().toString();
+  }
+
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    try {
+      int retryCount = lockConfiguration.getConfig().getInteger(LockConfiguration.LOCK_ACQUIRE_NUM_RETRIES_PROP_KEY);
+      int retryTimeout = lockConfiguration.getConfig().getInteger(LockConfiguration.LOCK_ACQUIRE_RETRY_WAIT_TIME_IN_MILLIS_PROP_KEY);
+      for (int i = 0; i < retryCount; i++) {
+        if (acquireLock()) {
+          LOG.info(generateLogStatement(LockState.ACQUIRED, generateLogSuffixString()));
+          return true;
+        } else {
+          Thread.sleep(retryTimeout);
+        }
+      }
+    } catch (InterruptedException e) {
+      throw new HoodieLockException("Failed to acquire file lock", e);
+    }
+    return false;
+  }
+
+  @Override
+  public void unlock() {
+    LOG.info(generateLogStatement(LockState.RELEASING, generateLogSuffixString()));
+    if (dataOutputStream != null) {
+      try {
+        dataOutputStream.close();
+        dataOutputStream = null;
+      } catch (IOException e) {
+        LOG.error(generateLogStatement(LockState.FAILED_TO_RELEASE, generateLogSuffixString()));
+        throw new HoodieLockException(generateLogStatement(LockState.FAILED_TO_RELEASE, generateLogSuffixString()), e);
+      }
+    }
+    LOG.info(generateLogStatement(LockState.RELEASED, generateLogSuffixString()));
+  }
+
+  private boolean acquireLock() {
+    if (lockDir == null || lockFile == null) {
+      return false;
+    }
+    try {
+      Path lockPath = new Path(lockDir);
+      Path lockFilePath = new Path(lockFile);
+      if (!fs.exists(lockPath)) {
+        fs.mkdirs(lockPath);
+      }
+      if (!fs.exists(lockFilePath)) {
+        // Pass the permissions during file creation itself
+        fs.create(lockFilePath, new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL),
+            false, fs.getConf().getInt("io.file.buffer.size", 4096),
+            fs.getDefaultReplication(lockFilePath), fs.getDefaultBlockSize(lockFilePath), null).close();
+        // set permission forcefully
+        fs.setPermission(lockFilePath, new FsPermission(FsAction.ALL, FsAction.ALL, FsAction.ALL));
+      }
+      dataOutputStream = fs.append(new Path(lockFile));
```

Comment:
At a high level, the `append` API is something that we cannot rely on. For example, append API is not supported for HDFS enabled with erasure-coding. Can we use a different API instead or come up with a lockFile created / check / deleted way given file creation and deletion in HDFS is atomic ?

Reply:
Thank you for your advice， i will try

Reply:
@n3nash 
I think it is ok to use append
1: EC needs to be opened manually， and ec is for special foleder not all hdfs file。
2：this lock support cleaning/expiring the lock after a failed
3.  i see carbondata also use file lock just like we do。  we have use this lock in product